lib/DhallToCabal/ConfigTree.hs

```diff
@@ -0,0 +1,172 @@
+{-# language DeriveFunctor #-}
+{-# language LambdaCase #-}
+{-# language OverloadedStrings #-}
+
+module DhallToCabal.ConfigTree ( ConfigTree(..), toConfigTree ) where
+
+import Control.Applicative
+import Control.Lens (deepOf)
+import Data.Functor.Identity
+import Data.Monoid
+import Dhall.Core hiding (Const)
+
+
+
+data ConfigTree cond a
+  = Leaf a
+  | Branch cond ( ConfigTree cond a ) ( ConfigTree cond a )
+  deriving (Functor, Show)
+
+
+
+toConfigTree :: ( Eq a, Eq s ) => Expr s a -> ConfigTree ( Expr s a ) ( Expr s a )
+toConfigTree e =
+  let
+    v =
+      "config"
+
+    saturated =
+      normalize (App e (Var v))
+
+    loop e =
+      let
+        Const condition =
+          deepOf
+            subExpr
+            ( findConfigUse v )
+            ( Const . First . Just ) e
+
+      in
+        case condition of
+          First Nothing ->
+            Leaf e
+
+          First (Just cond) ->
+            let
+              replaceWith x =
+                loop
+                  . normalize
+                  . runIdentity
+                  . deepOf
+                      subExpr
+                      ( exactly cond )
+                      ( const ( pure x ) )
+
+            in
+              Branch
+                cond
+                ( replaceWith ( BoolLit True ) e )
+                ( replaceWith ( BoolLit False ) e )
+
+  in loop saturated
+
+
+
+findConfigUse
+  :: Applicative f
+  => Var -> ( Expr s a -> f ( Expr s a ) ) -> Expr s a -> f ( Expr s a )
+findConfigUse x f e@(App (Field (Var x') "os") _) | x == x' = f e
+findConfigUse x f e@(App (Field (Var x') "arch") _) | x == x' = f e
+findConfigUse x f e@(App (App (Field (Var x') "impl") _) _) | x == x' = f e
+findConfigUse x f e@(App (Field (Var x') "flag") _) | x == x' = f e
+findConfigUse _ _ e = pure e
+
+
+
+exactly :: ( Applicative f, Eq a ) => a -> ( a -> f a ) -> a -> f a
+exactly x f e | x == e = f e
+exactly _ _ e = pure e
+
+
+
+subExpr
+  :: Applicative f
+  => ( Expr s a -> f ( Expr s a ) ) -> Expr s a -> f ( Expr s a )
+subExpr f = \case
```

Comment:
If you want I can derive `Data` for the `Expr` type so that you can use `subExpr = Control.Lens.uniplate`

Reply:
I think I might prefer this function itself added, so people can do traversals without needing `lens`. But if you don't want that, `Data` and `Generic` would be helpful additions.